Add support for imagesLoaded

jussikinnula opened the pull request:
Added support for imagesLoaded, and also tested it together with PR#17.

Testing can be done in an Angular 2 project with:

`npm install --save jussikinnula/angular2-masonry#testing`

jelgblad commented:
Hi! Have you tested this with both Webpack and SystemJS?
I've tried something similar to this before, but Webpack don't like referencing libraries that isn't installed. Even if require is placed within if-statement.
I don't remember which Webpack-version I tried this on, so I really don't know if this is the case anymore.

I don't want to force people to npm install imagesloaded if they're not using it.

jussikinnula commented:
Well, I'm personally using Webpack 2 and there's no trouble on having imagesloaded module referenced on angular2-masonry, as Webpack 2 can do ES6 tree shaking (of course you have to use module resolution ES6 to do that). With SystemJS it's not a problem just to include empty reference for imagesloaded, so it won't be loaded at all. Probably it would be then wise to add checks "if (this._imagesLoaded) {}" on all places imagesloaded is used, so if it would have empty reference then there wouldn't be any errors.

Alternative implementation would be separating the imagesloaded functionality to another module (lets call it "angular2-imagesloaded") which could be a container for anything that will have images loading on them. The module would give signal (with EventEmitter) to the child component when images are loaded, which would allow on our case running "this.layout()" everytime images were actually loaded.

However, the imagesloaded module is very small compared to anything on Angular 2 and even the Masonry module, so I wouldn't worry about having people to have yet another dependency on the first place. I guess the real improvement would be writing Masonry functionality on pure Angular 2 instead of wrapping the jQuery modules, both performance wise and making the foot print smaller. I've written pure masonry kind of functionality myself for Angular 1.x, but of course it's not directly "portable" to Angular 2.

I have updated all my pull requests based on the recent master, and updated the testing branch as well (with the pull request to upgrade to Angular 2.1.0 and new pull request to disable strict null checking, which has an open bug on Angular side).

jussikinnula commented:
Btw, a live demo of the latest combined test branch `npm install --save jussikinnula/angular2-masonry#testing` is live at https://www.angular.fi/ if you want to check how I use the module! :-)

ak0arjun commented:
Hi, 

  I am facing an issue on random browsing on using imagesLoaded fix.

DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
    at AngularMasonry.add (http://localhost:3000/node_modules/angular2-masonry/src/masonry.js:81:41)
    at AngularMasonryBrick.ngAfterViewInit (http://localhost:3000/node_modules/angular2-masonry/src/brick.js:22:22)
    at View_TribeComponent2.detectChangesInternal (/AppModule/TribeComponent/component.ngfactory.js:343:78)
    at View_TribeComponent2.AppView.detectChanges (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9355:18)
    at View_TribeComponent2.DebugAppView.detectChanges (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9448:48)
    at ViewContainer.detectChangesInNestedViews (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9520:41)
    at View_TribeComponent1.detectChangesInternal (/AppModule/TribeComponent/component.ngfactory.js:146:14)
    at View_TribeComponent1.AppView.detectChanges (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9355:18)
    at View_TribeComponent1.DebugAppView.detectChanges (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9448:48)
    at ViewContainer.detectChangesInNestedViews (http://localhost:3000/node_modules/@angular/core/bundles/core.umd.js:9520:41)

The worse part is it stops website working ?
Any suggestions

jussikinnula commented:
While trying to upgrade to later Angular 2 versions I did notice the same. The test repository is in the same branch still, but with updated Angular which is having this issue.

I did solve the problem actually on a site to be in production by not using imagesLoaded, but just adding a hook to angular2-masonry to trigger "images loaded" functionality externally. The mechanism is just triggered from `<img src="xxx.jpg" (load)="imageLoaded()">` which fires a ReplaySubject forward to angular2-masonry.

The patch is here:
https://github.com/jussikinnula/angular2-masonry/commit/c7ea13c3d190dacb2bed2f1fd805d2604dc6d6e2

And you can test it by:
`npm install --save jussikinnula/angular2-masonry#testing-universal`

Minimum code would be something like:

```
<masonry [trigger]="trigger">
  <masonry-brick>
    <img src="foo.jpg" (load)="imageLoaded()">
  </masonry-brick>
</masonry>
```

```
export class MyViewComponent {
  trigger: ReplaySubject<any> = new ReplaySubject(1);

  constructor() {}

  imageLoaded() {
    this.trigger.next(true);
  }
}
```

Besides the trouble you described, I also took care of making angular2-masonry to work on Angular Universal (e.g. server side rendering).

I'm not yet happy with the solution, it should be a bit more modular like a directive to be injected on all IMG elements inside Masonry. But I'm sharing the code just so that you can get further, while the better solution is yet to be implemented :-) ...

BR,
Jussi K.

jussikinnula commented:
I'm actually closing the pull request, since using imagesLoaded on Angular 2 (including Angular Universal) is not yet optimal.

ak0arjun commented:
Hi, 

I tried using the trigger fix in masonry.ts but my masonry.js is not getting updated with trigger variable, hence i am getting error masonry donot support trigger property.
Any suggestion how to regenerate masonry.js

jussikinnula commented:
You need to use my custom build (first remove the old one, `rm -rf node_modules/angular2-masonry`, and then install new `npm install --save jussikinnula/angular2-masonry#testing-universal`).

ak0arjun commented:
Ok, got it thanks! It worked :)
